File: backend/tournament.py
from __future__ import annotations

from typing import Dict, List, Mapping, Optional, Sequence, Set

from .constants import Result


PlayerDict = Mapping[str, object]


def _player_id(player: PlayerDict) -> int:
    return int(player["id"])  # type: ignore[index]


def gather_opponent_history(players: Sequence[PlayerDict], rounds: Sequence[Dict]) -> Dict[int, Set[int]]:
    history: Dict[int, Set[int]] = {_player_id(player): set() for player in players}
    for round_data in rounds:
        for pairing in round_data.get("pairings", []):
            player1 = pairing.get("player1")
            player2 = pairing.get("player2")
            if not player1 or not player2:
                continue
            history.setdefault(player1, set()).add(player2)
            history.setdefault(player2, set()).add(player1)
    return history


def recalculate_standings(players: Sequence[PlayerDict], rounds: Sequence[Dict]) -> List[Dict]:
    stats_map: Dict[int, Dict] = {}
    players_by_id = {_player_id(player): player for player in players}

    for player in players:
        player_id = _player_id(player)
        name = str(player.get("name", ""))
        full_name = player.get("fullName") or name
        add_score = float(player.get("addScore") or 0)
        stats_map[player_id] = {
            "id": player_id,
            "seed": int(player.get("seed", 0)),
            "name": full_name,
            "displayName": name,
            "addScore": add_score,
            "opponents": set(),
            "wins": 0,
            "losses": 0,
            "draws": 0,
            "byes": 0,
            "basePoints": 0.0,
        }

    for round_data in rounds:
        for pairing in round_data.get("pairings", []):
            player1 = pairing.get("player1")
            player2 = pairing.get("player2")
            result = pairing.get("result")

            if player1 not in stats_map:
                continue

            p1_stats = stats_map[player1]
            p2_stats = stats_map.get(player2) if player2 is not None else None

            if player2 is not None:
                p1_stats["opponents"].add(player2)
                if p2_stats is not None:
                    p2_stats["opponents"].add(player1)

            if result == Result.PLAYER1.value:
                p1_stats["wins"] += 1
                p1_stats["basePoints"] += 1
                if p2_stats is not None:
                    p2_stats["losses"] += 1
            elif result == Result.PLAYER2.value:
                if p2_stats is not None:
                    p2_stats["wins"] += 1
                    p2_stats["basePoints"] += 1
                p1_stats["losses"] += 1
            elif result == Result.DRAW.value:
                p1_stats["draws"] += 1
                p1_stats["basePoints"] += 0.5
                if p2_stats is not None:
                    p2_stats["draws"] += 1
                    p2_stats["basePoints"] += 0.5
            elif result == Result.BYE.value:
                p1_stats["byes"] += 1
                p1_stats["basePoints"] += 1
                p1_stats["opponents"].add("Bye")

    standings: List[Dict] = []
    for stats in stats_map.values():
        games_played = stats["wins"] + stats["losses"] + stats["draws"] + stats["byes"]
        total_points = stats["basePoints"] + stats["addScore"]
        win_percent = 0 if games_played == 0 else (stats["basePoints"] / games_played) * 100

        opponent_summaries: List[str] = []
        for opponent in stats["opponents"]:
            if opponent == "Bye":
                opponent_summaries.append("Bye")
                continue
            opponent_data = players_by_id.get(opponent)
            if opponent_data is not None:
                opponent_name = opponent_data.get("name") or opponent_data.get("fullName") or "Unknown"
                opponent_summaries.append(f"{opponent_name} (#{opponent})")
            else:
                opponent_summaries.append(f"#{opponent}")

        standings.append(
            {
                **stats,
                "gamesPlayed": games_played,
                "totalPoints": total_points,
                "winPercent": win_percent,
                "opponentSummaries": opponent_summaries,
            }
        )

    standings.sort(
        key=lambda entry: (
            -entry["totalPoints"],
            -entry["basePoints"],
            -entry["winPercent"],
            entry["seed"],
        )
    )
    return standings


def can_generate_next_round(rounds: Sequence[Dict]) -> bool:
    if not rounds:
        return False
    return all(
        pairing.get("result") != Result.UNPLAYED.value
        for round_data in rounds
        for pairing in round_data.get("pairings", [])
    )


def create_swiss_pairings(player_ids: Sequence[int], opponent_history: Dict[int, Set[int]]) -> List[List[int]]:
    pool = list(player_ids)
    bye_player: Optional[int] = None

    if len(pool) % 2 == 1:
        bye_player = pool.pop()

    solution: Optional[List[List[int]]] = None

    def backtrack(available: List[int], current_pairs: List[List[int]]) -> bool:
        nonlocal solution
        if not available:
            solution = list(current_pairs)
            return True

        player = available[0]
        rest = available[1:]

        for index, opponent in enumerate(rest):
            if opponent in opponent_history.get(player, set()):
                continue
            remaining = rest[:index] + rest[index + 1 :]
            if backtrack(remaining, current_pairs + [[player, opponent]]):
                return True

        for index, opponent in enumerate(rest):
            remaining = rest[:index] + rest[index + 1 :]
            if backtrack(remaining, current_pairs + [[player, opponent]]):
                return True

        return False

    if not backtrack(pool, []):
        raise ValueError("Unable to create Swiss pairings without conflicts.")

    final_pairs = list(solution or [])
    if bye_player is not None:
        final_pairs.append([bye_player])
    return final_pairs

File: shared/tournament.js
import { RESULT } from './constants.js';

export function gatherOpponentHistory(players, rounds) {
  const history = new Map(players.map((player) => [player.id, new Set()]));

  rounds.forEach((round) => {
    round.pairings.forEach((pairing) => {
      if (!pairing.player1 || !pairing.player2) return;
      history.get(pairing.player1)?.add(pairing.player2);
      history.get(pairing.player2)?.add(pairing.player1);
    });
  });

  return history;
}

export function recalculateStandings(players, rounds) {
  const statsMap = new Map();
  const playersById = new Map(players.map((player) => [player.id, player]));

  players.forEach((player) => {
    statsMap.set(player.id, {
      id: player.id,
      seed: player.seed,
      name: player.fullName || player.name,
      displayName: player.name,
      addScore: player.addScore ?? 0,
      opponents: new Set(),
      wins: 0,
      losses: 0,
      draws: 0,
      byes: 0,
      basePoints: 0
    });
  });

  rounds.forEach((round) => {
    round.pairings.forEach((pairing) => {
      const { player1, player2, result } = pairing;
      if (!statsMap.has(player1)) return;
      const p1Stats = statsMap.get(player1);
      const p2Stats = player2 ? statsMap.get(player2) : null;

      if (player2) {
        p1Stats.opponents.add(player2);
        if (p2Stats) {
          p2Stats.opponents.add(player1);
        }
      }

      switch (result) {
        case RESULT.PLAYER1:
          p1Stats.wins += 1;
          p1Stats.basePoints += 1;
          if (p2Stats) {
            p2Stats.losses += 1;
          }
          break;
        case RESULT.PLAYER2:
          if (p2Stats) {
            p2Stats.wins += 1;
            p2Stats.basePoints += 1;
          }
          p1Stats.losses += 1;
          break;
        case RESULT.DRAW:
          p1Stats.draws += 1;
          p1Stats.basePoints += 0.5;
          if (p2Stats) {
            p2Stats.draws += 1;
            p2Stats.basePoints += 0.5;
          }
          break;
        case RESULT.BYE:
          p1Stats.byes += 1;
          p1Stats.basePoints += 1;
          p1Stats.opponents.add('Bye');
          break;
        default:
          break;
      }
    });
  });

  const standings = Array.from(statsMap.values()).map((stats) => {
    const gamesPlayed = stats.wins + stats.losses + stats.draws + stats.byes;
    const totalPoints = stats.basePoints + stats.addScore;
    const winPercent = gamesPlayed === 0 ? 0 : (stats.basePoints / gamesPlayed) * 100;
    const opponentSummaries = Array.from(stats.opponents).map((opponent) => {
      if (opponent === 'Bye') {
        return 'Bye';
      }
      const opponentData = playersById.get(opponent);
      return opponentData ? `${opponentData.name} (#${opponentData.id})` : `#${opponent}`;
    });

    return {
      ...stats,
      gamesPlayed,
      totalPoints,
      winPercent,
      opponentSummaries
    };
  });

  standings.sort((a, b) => {
    if (b.totalPoints !== a.totalPoints) {
      return b.totalPoints - a.totalPoints;
    }
    if (b.basePoints !== a.basePoints) {
      return b.basePoints - a.basePoints;
    }
    if (b.winPercent !== a.winPercent) {
      return b.winPercent - a.winPercent;
    }
    return a.seed - b.seed;
  });

  return standings;
}

export function canGenerateNextRound(rounds) {
  if (!rounds.length) return false;
  return rounds.every((round) =>
    round.pairings.every((pairing) => pairing.result !== RESULT.UNPLAYED)
  );
}

export function createSwissPairings(playerIds, opponentHistory) {
  const pool = [...playerIds];
  let byePlayer = null;

  if (pool.length % 2 === 1) {
    byePlayer = pool.pop();
  }

  let solution = null;

  function backtrack(available, currentPairs) {
    if (available.length === 0) {
      solution = currentPairs;
      return true;
    }

    const [player, ...rest] = available;

    for (let i = 0; i < rest.length; i += 1) {
      const opponent = rest[i];
      if (opponentHistory.get(player)?.has(opponent)) {
        continue;
      }
      const remaining = rest.slice(0, i).concat(rest.slice(i + 1));
      if (backtrack(remaining, [...currentPairs, [player, opponent]])) {
        return true;
      }
    }

    for (let i = 0; i < rest.length; i += 1) {
      const opponent = rest[i];
      const remaining = rest.slice(0, i).concat(rest.slice(i + 1));
      if (backtrack(remaining, [...currentPairs, [player, opponent]])) {
        return true;
      }
    }

    return false;
  }

  if (!backtrack(pool, [])) {
    throw new Error('Unable to create Swiss pairings without conflicts.');
  }

  const finalPairs = solution ? [...solution] : [];
  if (byePlayer !== null) {
    finalPairs.push([byePlayer]);
  }
  return finalPairs;
}
